File: boundaries.py
'''
 @file      material.py
 @brief     Contains the Boundaries class
 @author    Geoffrey Gunow
 @date      October 7, 2015
'''

'''
 @class Boundaries boundaries.py "boundaries.py"
 @brief Contains information about the boundaries of a geometry and the
        associated boundary conditions
'''
class Boundaries():
    def __init__(self, x_min, x_max, y_min, y_max, z_min, z_max):

        # create dictionary to store boundaries
        self._surfaces = dict()
        for surface in ['x', 'y', 'z']:
            self._surfaces[surface] = {'min':{}, 'max':{}}

        # store boundary locations
        self._surfaces['x']['min']['coord'] = x_min
        self._surfaces['x']['max']['coord'] = x_max
        self._surfaces['y']['min']['coord'] = y_min
        self._surfaces['y']['max']['coord'] = y_max
        self._surfaces['z']['min']['coord'] = z_min
        self._surfaces['z']['max']['coord'] = z_max
        
        # store boundary types (0 = vacuum, 1 = reflective)
        for surface in self._surfaces:
            for side in ['min', 'max']:
                self._surfaces[surface][side]['type'] = 1 

    '''
     @brief returns the coordinate of the maximum or minimum boundary for a
            given axis
    '''
    def get_surface_coord(self, axis, side):
        return self._surfaces[axis][side]['coord']

    '''
     @brief returns the material type for a given boundary
    '''
    def get_surface_type(self, axis, side):
        return self._surfaces[axis][side]['type']

File: Boundaries.cpp
/* 
 @file      Boundaries.cpp
 @brief     contains functions for the Boundaries class
 @author    Luke Eure
 @date      January 9 2016
*/

#include "Boundaries.h"

/*
 @brief     constructor for Boundaries class
*/
Boundaries::Boundaries() {}

/*
 @brief     deconstructor
*/
Boundaries::~Boundaries() {}

/*
 @brief     set the surfaces into the geometry
 @paraam    axis 0, 1, or 2 corresponding to x y and z
 @param     side 0 or 1 corresponding to the minimum or maximum of the geometry
*/
void Boundaries::setSurface(Axes axis, min_max side, Surface* surface) {
    _surfaces[2*axis + side] = surface;
}

/*
 @brief     return the position of the surface
 @paraam    axis 0, 1, or 2 corresponding to x y and z
 @param     side 0 or 1 corresponding to the minimum or maximum of the geometry
 @return    the position of the surface within the geometry
*/
float Boundaries::getSurfaceCoord(int axis, int side) {
    return _surfaces[axis*2+side]->getPosition();
}

/*
 @brief     return the type of a surface
 @paraam    axis 0, 1, or 2 corresponding to x y and z
 @param     side 0 or 1 corresponding to the minimum or maximum of the geometry
 @return    the type of the surface, 0 = vacuum 1 = reflective
*/
BoundaryType Boundaries::getSurfaceType(int axis, int side) {
    return _surfaces[axis*2+side]->getType();
}

/*
 @brief     function that samples a random location within a bounding box.
 @details   a point is randomly and uniformally sampled in the bounding box 
            provided in the input.
 @param     bounds a Boundaries object containing the limits of the
            bounding box
 @return    _dist_location a vector that contains the coordinates of a point
*/
std::vector <double> Boundaries::sampleLocation(Neutron* neutron) {
    std::vector <double> _dist_location(3);
    for (int axis=0; axis<3; ++axis) {
        double width = getSurfaceCoord(axis, MAX) - getSurfaceCoord(axis, MIN);
        double coord = getSurfaceCoord(axis, MIN) + width * neutron->arand();
        _dist_location[axis] = coord;
    }
    return _dist_location;
}
